### DISCUSSION TITLE
Logic library, defining interrupt handler with ISR(CCL_CCL_vect)

### DISCUSSION THREAD

**Discussion by alx2009** [SpenceKonde/DxCore#418] [Q&A] [answered]
Reading the documentation for the Logic library, under "Advanced Interrupt" it says that "New in 1.5.0, it is possible to use interrupts "manually" while still using the Logic library". Essentially that means defined the interrupt with ISR(CCL_CCL_vect) rather than using attachInterrupt.

However, when I try I get a compilation error. Looking at the source code for LogicISR.cpp I understand for this to work library.properties should include dot_a_linkage=true. This is not the case, hence the conflict detected by the linker.

Now, this is easily fixed adding a line dot_a_linkage=true to library.properties.

Is there any reason this is not done in DxCore?

**Comment by SpenceKonde** [comment]
AAAH! Someone who understands what "dot_a_linkage = true" does! Explain it to me! I thought it was just for precompiled libraries. 

I concluded that there's no way it could be required, because HardwareSerial relies upon the fact that if no function in a compilation unit is referenced, nothing from the compilation unit goes into the binary. So why if we're not referencing anything in LogicISR, would we need dot_a_linkage?

**Reply by alx2009** [reply]
Well, I think the Arduino platform specification explains what dot_a_linkage does: in short the library object files are passed as one archive (.a file) rather than one by one (as .o files). 
The big question is: does it matter? I am not an expert of gcc and its linker, but from what I understand, all .o files are linked to produce the executable. If the linker finds a duplicated symbol, that is a fatal error. However, when an .o file is part of an archive it's only considered in case the linker need at least one symbol defined in that file (and said symbol is not found in the standalone .o files). 

I believe the linker does optimize away what is not needed, but only as a second step, once it's happy with all the input.

I suspect the reason you don't need dot_a_linkage = true for Serial is that the latter is part of the core rather than a library.  I am not 100% sure, but I think the core is always passed as an archive rather than a collection of object files.

**Reply by SpenceKonde** [reply]
Huh. never knew that. I didn't use it because as I understood it, the options was a do-nothing option since the library wasn't precompiled. Hence it was unnecessary. Can you confirm that adding that to the library properties fixes it?

**Reply by alx2009** [reply]
I agree there is very little in terms of explanations of what dot_a_linkage does! It is needed for precompile libraries, and this must be the main use. Otherwise it's a do nothing - except when it is exploited by the library like in this case.
I can confirm that adding the option it compiles and (more important for me) it actually works. I also save about 200 bytes of flash by not using attachInterrupt.

Just for future reference, I have tested with the application I am developing, here: https://github.com/alx2009/K197Display

The Main branch uses attachInterrupt (file K197PushButtons.cpp). It works with or without the  dot_a_linkage in library.properties (we would not want to break that, right?)

The Optimize-CCL-interrupt branch uses ISR(CCL_CCL_vect). Without dot_a_linkage the compilation fails with the following error:

C:\Users\dax\AppData\Local\Temp\arduino_build_900592\libraries\Logic\LogicISR.cpp.o (symbol from plugin): In function `Logic::attachInterrupt(void (*)(), unsigned char)':
(.text+0x0): multiple definition of `__vector_7'
C:\Users\dax\AppData\Local\Temp\arduino_build_900592\sketch\K197PushButtons.cpp.o (symbol from plugin):(.text+0x0): first defined here
c:/users/dax/appdata/local/arduino15/packages/dxcore/tools/avr-gcc/7.3.0-atmel3.6.1-azduino6/bin/../lib/gcc/avr/7.3.0/../../../../avr/bin/ld.exe: Disabling relaxation: it will not work with multiple definitions

And the compilation fails.

Adding the line dot_a_linkage=true in library.properties it compiles & everything works the same (well, latency of the interrupt should improve of course)

**Comment by SpenceKonde** [comment, marked as answer]
Thanks. I've checked in fixes for this in logic and comparator, which I think are the only libraries where this would crop up.
